orc/src/test/java/org/apache/iceberg/orc/TestORCSchemaUtil.java

```diff
@@ -0,0 +1,207 @@
+    assertEquals(3, nestedCol.findSubtype("c").getId());
+    assertEquals(TypeDescription.Category.DATE, nestedCol.findSubtype("c").getCategory());
+  }
+
+  @Test
+  public void testSchemaEvolutionNested() {
+    Types.StructType nestedStructType = Types.StructType.of(
+        optional(2, "b", Types.StringType.get()),
+        optional(3, "c", Types.DateType.get())
+    );
+    Schema originalSchema = new Schema(
+        optional(1, "a", nestedStructType)
+    );
+
+    // Original mapping (stored in ORC)
+    TypeDescription orcSchema = ORCSchemaUtil.toOrc(originalSchema);
+
+    // Evolve schema
+    Types.StructType newNestedStructType = Types.StructType.of(
+        optional(3, "cc", Types.DateType.get()),
+        optional(2, "bb", Types.StringType.get())
+    );
+    Schema evolveSchema = new Schema(
+        optional(1, "aa", newNestedStructType)
+    );
+
+    TypeDescription newOrcSchema = ORCSchemaUtil.toOrc(evolveSchema, orcSchema);
+    assertEquals(1, newOrcSchema.getChildren().size());
+    assertEquals(TypeDescription.Category.STRUCT, newOrcSchema.findSubtype("a").getCategory());
+    TypeDescription nestedCol = newOrcSchema.findSubtype("a");
+    assertEquals(2, nestedCol.findSubtype("c").getId());
+    assertEquals(TypeDescription.Category.DATE, nestedCol.findSubtype("c").getCategory());
+    assertEquals(3, nestedCol.findSubtype("b").getId());
+    assertEquals(TypeDescription.Category.STRING, nestedCol.findSubtype("b").getCategory());
+  }
+
+  @Test
+  public void testTypePromotions() {
+    Schema originalSchema = new Schema(
+        optional(1, "a", Types.IntegerType.get()),
+        optional(2, "b", Types.FloatType.get()),
+        optional(3, "c", Types.DecimalType.of(10, 2))
+    );
+
+    // Original mapping (stored in ORC)
+    TypeDescription orcSchema = ORCSchemaUtil.toOrc(originalSchema);
+
+    // Evolve schema
+    Schema evolveSchema = new Schema(
+        optional(1, "a", Types.LongType.get()),
+        optional(2, "b", Types.DoubleType.get()),
+        optional(3, "c", Types.DecimalType.of(15, 2))
+    );
+
+    TypeDescription newOrcSchema = ORCSchemaUtil.toOrc(evolveSchema, orcSchema);
+    assertEquals(3, newOrcSchema.getChildren().size());
+    assertEquals(1, newOrcSchema.findSubtype("a").getId());
+    assertEquals(TypeDescription.Category.LONG, newOrcSchema.findSubtype("a").getCategory());
+    assertEquals(2, newOrcSchema.findSubtype("b").getId());
+    assertEquals(TypeDescription.Category.DOUBLE, newOrcSchema.findSubtype("b").getCategory());
+    TypeDescription decimalC = newOrcSchema.findSubtype("c");
+    assertEquals(3, decimalC.getId());
+    assertEquals(TypeDescription.Category.DECIMAL, decimalC.getCategory());
+    assertEquals(15, decimalC.getPrecision());
+    assertEquals(2, decimalC.getScale());
+  }
+
+  @Test(expected = IllegalArgumentException.class)
+  public void testInvalidTypePromotions() {
```

Comment:
I recommend using `AssertHelpers.assertThrows` from iceberg-core instead of `expected`. It's easier to check the error message and you can have more than one assertion per test.